jacdac-ble/jdble.cpp

```diff
@@ -0,0 +1,150 @@
+#include "pxt.h"
+#include "jdlow.h"
+#include "mbbridge.h"
+#include "JacdacBLE.h"
+
+#ifdef MICROBIT_CODAL
+
+#if CONFIG_ENABLED(DEVICE_BLE)
+
+namespace pxt {
+void (*logJDFrame)(const uint8_t *data);
+void (*sendJDFrame)(const uint8_t *data);
+uint64_t current_time_us() {
+    return system_timer_current_time_us();
+}
+} // namespace pxt
+
+namespace jacdac {
+
+
+static JacdacBLE* jdble;
+
+// from device to PC
+void jdble_tx(const uint8_t *data) {
+    jdble->send((uint8_t *)data, JD_FRAME_SIZE(((jd_frame_t*)data)));
+}
+// from PC to device.
+void jdble_rx(MicroBitEvent) {
+    pxt::sendJDFrame((const uint8_t *)(void *)jdble->read().getBytes());
+}
+
+void jdble_init() {
+    jdble = new JacdacBLE(*uBit.ble);
+    uBit.messageBus.listen(DEVICE_ID_JACDAC_BLE, MICROBIT_JACDAC_S_EVT_RX, jdble_rx, MESSAGE_BUS_LISTENER_IMMEDIATE);
+    pxt::logJDFrame = jdble_tx;
```

Comment:
this will disable WebUSB, is this expected?

Reply:
I don't know what to do here... if we want to power a micro:bit over whilst BLE is enabled, we could end up creating a loop of death if jacdac-ts decides to connect to both :smile:... Also, I haven't tested this, but I suspect softdevice will get upset if the mcu is in debug mode with an active debugging connection.

Thoughts?

Reply:
For a start I think this is fine. We can refine later.